patches/api/0009-Per-Entity-Tracking-Ranges.patch

```diff
@@ -0,0 +1,108 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: fluse1367 <3984402-fluse1367@users.noreply.gitlab.com>
+Date: Tue, 19 Jan 2021 11:53:21 +0100
+Subject: [PATCH] Per Entity Tracking Ranges
+
+
+diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
+index dc7e9983b89726625acce95026b45695b3f387b6..c6c999f27d96b52981204a56c2ad8eb0869c0df4 100644
+--- a/src/main/java/org/bukkit/entity/Entity.java
++++ b/src/main/java/org/bukkit/entity/Entity.java
+@@ -719,5 +719,23 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
+     default de.tr7zw.changeme.nbtapi.NBTCompound getNBTC() {
+         return getNBT().getPersistentDataContainer();
+     }
++
++    /**
++     * Returns the tracking (render) distance for this specific {@link Entity},<br>
++     *     falls back to {@link EntityType#getTrackingDistance(World)} if not set.
++     * @return the tracking (render) distance for this specific {@link Entity}.
++     */
++    int getTrackingDistance();
++
++    /**
++     * the tracking (render) distance for this specific {@link Entity}.
++     * @param distance the tracking distance to set.
++     */
++    void setTrackingDistance(int distance);
++
++    /**
++     * Sets the tracking (render) distance for this specific {@link Entity} back to default.
++     */
++    void resetTrackingDistance();
+     // Yatopia end
+ }
+diff --git a/src/main/java/org/bukkit/entity/EntityType.java b/src/main/java/org/bukkit/entity/EntityType.java
+index 692b75eb78405874077c850bfc72e247ccc80860..3def6f19cdd3acf51c1ca96385e7c78c3a392e2f 100644
+--- a/src/main/java/org/bukkit/entity/EntityType.java
++++ b/src/main/java/org/bukkit/entity/EntityType.java
+@@ -287,6 +287,8 @@ public enum EntityType implements Keyed {
+     private final short typeId;
+     private final boolean independent, living;
+     private final NamespacedKey key;
++    private int trackingDistance = -1; // Yatopia
++    private final Map<World, Integer> trackingDistances = new HashMap<>(); // Yatopia
```

Comment:
Maybe directly use Fastutil? We have it so why not use a very small performance enhancement :)

Reply:
Not always. Fastutil and similar alternatives to Java's built in HashSet, HashMap, and other collections are primarily designed with reducing memory footprint in mind. Although in some instances they offer performance gains varying in size a majority of the time, especially with HashMap alternatives they greatly reduce performance in most cases. I suggest you take a look at this paper: https://research.spec.org/icpe_proceedings/2017/proceedings/p389.pdf